src/bin/cross-util.rs

```diff
@@ -0,0 +1,193 @@
+#![deny(missing_debug_implementations, rust_2018_idioms)]
+
+use std::path::{Path, PathBuf};
+use std::process::Command;
+
+use clap::{Parser, Subcommand};
+use cross::CommandExt;
+
+// known image prefixes, with their registry
+// the docker.io registry can also be implicit
+const GHCR_IO: &str = "ghcr.io/cross-rs/";
+const RUST_EMBEDDED: &str = "rustembedded/cross:";
+const DOCKER_IO: &str = "docker.io/rustembedded/cross:";
+const IMAGE_PREFIXES: &[&str] = &[GHCR_IO, DOCKER_IO, RUST_EMBEDDED];
+
+#[derive(Parser, Debug)]
+#[clap(version, about, long_about = None)]
+struct Cli {
+    #[clap(subcommand)]
+    command: Commands,
+}
+
+#[derive(Subcommand, Debug)]
+enum Commands {
+    /// List cross images in local storage.
+    ListImages {
+        /// Provide verbose diagnostic output.
+        #[clap(short, long)]
+        verbose: bool,
+        /// Container engine (such as docker or podman).
+        #[clap(long)]
+        engine: Option<String>,
+    },
+    /// Remove cross images in local storage.
+    RemoveImages {
+        /// If not provided, remove all images.
+        targets: Vec<String>,
+        /// Remove images matching provided targets.
+        #[clap(short, long)]
+        verbose: bool,
+        /// Force removal of images.
+        #[clap(short, long)]
+        force: bool,
+        /// Remove local (development) images.
+        #[clap(short, long)]
+        local: bool,
+        /// Container engine (such as docker or podman).
+        #[clap(long)]
+        engine: Option<String>,
+    },
+}
+
+fn get_container_engine(engine: &Option<String>) -> Result<PathBuf, which::Error> {
+    if let Some(ce) = engine {
+        which::which(ce)
+    } else {
+        cross::get_container_engine()
+    }
+}
+
+fn is_cross_image(image: &str) -> bool {
+    IMAGE_PREFIXES.iter().any(|i| image.starts_with(i))
+}
+
+fn is_local_image(image: &str) -> bool {
+    // this cannot panic: we've formatted our image as ${repo}:${tag}
+    let tag = image.split_once(':').unwrap().1;
+    tag.starts_with("local")
+}
+
+fn get_cross_images(engine: &Path, verbose: bool, local: bool) -> cross::Result<Vec<String>> {
+    let stdout = Command::new(engine)
+        .arg("images")
+        .arg("--format")
+        .arg("{{.Repository}}:{{.Tag}}")
+        .run_and_get_stdout(verbose)?;
+
+    let mut images: Vec<String> = stdout
+        .lines()
+        .filter(|image| is_cross_image(image))
+        .filter(|image| local || !is_local_image(image))
+        .map(|image| image.to_string())
+        .collect();
+    images.sort();
+
+    Ok(images)
+}
+
+// the old rustembedded targets had the following format:
+//  repository = (${registry}/)?rustembedded/cross
+//  tag = ${target}(-${version})?
+// our target triple must match `[A-Za-z0-9_-]`
+fn rustembedded_target(tag: &str) -> Option<String> {
```

Comment:
why does this return an option?

Reply:
It should not. I had originally planned some checking if the triple had <3 components, but I don't think that's a great idea anymore (say, a custom image someone also wants to remove).